### ISSUE TITLE
[MINVOKER-323] DefaultVersionResolver is inflicting ArtifactNotFoundException for classifiers with SNAPSHOT version

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-invoker-plugin#559]
**[Tuomas Kiviaho](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tuomas_kiviaho)** opened **[MINVOKER-323](https://issues.apache.org/jira/browse/MINVOKER-323?redirect=false)** and commented

I use classifier artifact along with the artifact itself as a dependency in a Maven Invoker Plugin project. The project that calls the invoker has the artifact itself as a dependency, but no reference to the classifier.

This causes resolving of the maven-metadata.xml for the project itself plus downloading of the dependency artifact.When invoker is called the artifact is already downloaded to the local repo and cached with SNAPSHOT key.

This causes the DefaultVersionResolver to merge the version information of the SNAPSHOT\<myclassifier>:jar artifact that is now being resolved with the already downloaded SNAPSHOT key. Since the local version is newer than the repo version the DefaultVersionResolver thinks SNAPSHOT\<myclassifier>:jar to be outdated thus overriding it with local repo.

Since the SNAPSHOT\<myclassifier>:jar doesn't exist in the local repo there are no remote report left to try the DefaultArtifactResolver fails ultimately to ArtifactNotFoundException since there was no download task.

```java
[INFO] [DEBUG] Resolving artifact <mygroupid>.<myartifactid>:jar:<myclassifier>:<mybaseversion>-SNAPSHOT from []
```

 

---

**Issue Links:**
- [MINVOKER-313](https://issues.apache.org/jira/browse/MINVOKER-313) Get rid of maven-artifact-transfer

- [MNG-7548](https://issues.apache.org/jira/browse/MNG-7548) Kill off "legacy" repository metadata support

**Comment by jira-importer** [comment]
**[Tuomas Kiviaho](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tuomas_kiviaho)** commented

The problem is that even though metadata's `<lastUpdated>` is checked the existence of corresponding `<snapshotVersion><classifier>` of `SNAPSHOT<classifier>:<extension>` is not assured to exist in the repo that `SNAPSHOT` points to.

```java
String key = SNAPSHOT + getKey( artifact.getClassifier(), artifact.getExtension() );
merge( infos, SNAPSHOT, key );
if ( !resolve( result, infos, key ) )
{
    result.setVersion( version );
}
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Are you certain that this is done by Resolver and not Maven? Maven updates metadata sometimes as well.

**Comment by jira-importer** [comment]
**[Tuomas Kiviaho](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tuomas_kiviaho)** commented

> Are you certain that this is done by Resolver and not Maven? Maven updates metadata sometimes as well.

I'm not sure if I see the relevancy in this case the metadata is ok to begin with and Maven just hands over the task to repository system and it's done in one sweep. Technically parts of the resolver seem to be spread to the middle ground in form of the maven-resolver provider if this question is about where the issue should have been opened.

**Comment by jira-importer** [comment]
**[Tuomas Kiviaho](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tuomas_kiviaho)** commented

Here are the metadatas to give a clearer picture of the local repo state just prior to invoker:integration-test goal execution. It seems that invoker:install goal generates (partial???) `maven-metadata-local.xml` that I may have mistaken as the actual local repository while debugging since they both point to the same location in my settings. And for sure that one is missing references to pom and other artifacts.

But still it's the weird aftermath that `DefaultVersionResolver::resolveVersion` applies when the metadataresults have been already merged together successfully. It can't be just an optimization to see if the metadata of the local repo is newer than the remote since that's done already so have no idea what the code snippet in my first comment is all about.

This logic is ultimately causing the failure and invoker is just a contributor to the cause.

As a work-a-round I need to start finding an alternative/patch for invoker:install for the time being, but it would be nice to get clarification on what is going on with the version resolver post processing the snapshot classifiers repo in such way.

```xml
<metadata>
  <groupId>mygroupid</groupId>
  <artifactId>myartifactid</artifactId>
  <version>x.y.z-SNAPSHOT</version>
  <versioning>
    <snapshot>
      <localCopy>true</localCopy>
    </snapshot>
    <lastUpdated>20211117135742</lastUpdated>
  </versioning>
</metadata>
```

```xml
<?xml version="1.0" encoding="UTF-8"?>
<metadata modelVersion="1.1.0">
  <groupId>mygroupid</groupId>
  <artifactId>myartifactid</artifactId>
  <version>x.y.z-SNAPSHOT</version>
  <versioning>
    <snapshot>
      <timestamp>20211112.112753</timestamp>
      <buildNumber>2</buildNumber>
    </snapshot>
    <lastUpdated>20211112112753</lastUpdated>
    <snapshotVersions>
      <snapshotVersion>
        <extension>jar</extension>
        <value>x.y.z-20211112.112753-2</value>
        <updated>20211112112753</updated>
      </snapshotVersion>
      <snapshotVersion>
        <extension>pom</extension>
        <value>x.y.z-20211112.112753-2</value>
        <updated>20211112112753</updated>
      </snapshotVersion>
      <snapshotVersion>
        <classifier>myclassifier</classifier>
        <extension>jar</extension>
        <value>x.y.z-20211112.112753-2</value>
        <updated>20211112112753</updated>
      </snapshotVersion>
      <snapshotVersion>
        <classifier>sources</classifier>
        <extension>jar</extension>
        <value>x.y.z-20211112.112753-2</value>
        <updated>20211112112753</updated>
      </snapshotVersion>
    </snapshotVersions>
  </versioning>
</metadata>
```

**Comment by jira-importer** [comment]
**[Tuomas Kiviaho](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tuomas_kiviaho)** commented

I found a reference to this oddity via blame. MNG-4955 states that this is a remnant from fixing the early days of metadata files (MNG-4452).  It's still unclear to me what this exactly fixed even though there is a clear explanation for it.

I can't fathom why it wouldn't have without this fix. Maybe if the metadata for `SNAPSHOT<myclassifier>:jar` was missing all-together then falling back to whatever repo with the latest `SNAPSHOT` key would function as a fallback, but that's not the described use case.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Tuomas Kiviaho](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tuomas_kiviaho), can you provide a sample project for this?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Tuomas Kiviaho](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tuomas_kiviaho), let me know whether you can provide a sample. That would make the investigation much simpler.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Still waiting for a sample.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Still waiting...

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Yes, MNG-4452 introduced new maven metadata XML model version "1.1.0", as original (used by Maven2) has this limitation from beginning. Moreover, some MRMs like Nexus transform the served XML based on UA of the client (detects Maven2 vs Maven3) UNLESS cached/deployed Maven metadata is v1.0.0...

It has to be examined how 1.0.0 version model XML got in picture at all...

**Comment by jira-importer** [comment]
**[Tuomas Kiviaho](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tuomas_kiviaho)** commented

Thanks for the clarification.

My workaround was to switch from `invoker:install` to premature installing at `pre-integration-test` phase. If I recall correctly I stopped poking around when I saw that invoker did use shared metadata creation but had it's own rather crude implementation for it. Install plugin on the other hand didn't  "corrupt" the local repo with such partial (this is the model 1.0.0 I presume) metadata and thus I was able to continue without too much of a fuzz.

I'd close the issue by knowing now that the limitation had left it's marks/scars to the codebase.

PS. IMO `invoker:install` might be worth looking at with pair of eyes that know more about  the subtleties involved with metadata generation and resolving.
